### PR TITLE
Add `vec_ptype2()` and `vec_cast()` methods for table objects

### LINKED ISSUE #904: support for "table"
originally https://github.com/tidyverse/dplyr/issues/4963

``` r
vctrs::vec_c(table("tennis"))
#> Error: Can't find vctrs or base methods for concatenation.
#> vctrs methods must be implemented for class `table`.
#> See <https://vctrs.r-lib.org/articles/s3-vector.html>.
```

<sup>Created on 2020-03-10 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes #904 
Closes tidyverse/dplyr#4963

This roughly follows how ptype2 and cast methods have been implemented for glue. 
https://github.com/tidyverse/glue/blob/master/R/vctrs.R

The example from tidyverse/dplyr#4963:

``` r
library(dplyr)

ex <- table(iris$Species) %>%
  expand.grid() 

ex
#>   Var1
#> 1   50
#> 2   50
#> 3   50

class(ex$Var1)
#> [1] "table"

ex <- mutate(ex, Var2 = Var1)

ex
#>   Var1 Var2
#> 1   50   50
#> 2   50   50
#> 3   50   50

class(ex$Var2)
#> [1] "table"
```

<sup>Created on 2020-03-11 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

We do have at least one weird edge case. With `vec_rbind()`, we call `as_df_row()` on each input. For array-like objects with a dimensionality of 2, this calls `as.data.frame()`. But there is an `as.data.frame.table()` method that does something different from what it does for arrays. It doesn't error, it's just odd and incorrect, but I think `vec_rbind()` will have to be tweaked in a separate PR to fix it.

``` r
library(vctrs)

x <- vctrs:::new_table(1:4, c(2L, 2L))
dimnames(x) <- list(c("r1", "r2"), c("c1", "c2"))

x
#>    c1 c2
#> r1  1  3
#> r2  2  4

as.data.frame(x)
#>   Var1 Var2 Freq
#> 1   r1   c1    1
#> 2   r2   c1    2
#> 3   r1   c2    3
#> 4   r2   c2    4

vec_rbind(x, x)
#>   c1 c2 NA
#> 1 r1 c1  1
#> 2 r2 c1  2
#> 3 r1 c2  3
#> 4 r2 c2  4
#> 5 r1 c1  1
#> 6 r2 c1  2
#> 7 r1 c2  3
#> 8 r2 c2  4
```

### CODE REVIEW

**Comment by DavisVaughan** [comment]
Apparently on R 3.2 there was either no `[.table` method or it didn't keep the table class, so we have a failure there. I'll tweak the test.

R 3.2

```r
> x <- table(1, 1)
> 
> class(x)
[1] "table"
> 
> class(x[1, 1, drop = FALSE])
[1] "matrix"
```

R 3.6

```r
x <- table(1, 1)

class(x)
#> [1] "table"

class(x[1, 1, drop = FALSE])
#> [1] "table"
```

File: R/type-table.R [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  } else {
+    vec_default_cast(x, to, ..., x_arg = x_arg, to_arg = to_arg)
+  }
+}
+
+# ------------------------------------------------------------------------------
+
+new_table <- function(x = integer(), dim = 0L) {
+  # `table()` doesn't support long vectors, even though `tabulate()` does
+  if (!is.integer(x)) {
```
Comment (lionel-) [inline R/type-table.R]:
```suggestion
  if (!is_integer(x)) {
```

This way it's quite clear we are not dispatching (e.g. `is.character()` dispatches).

**Comment by romainfrancois** [comment]
@DavisVaughan still seeing this: 

``` r
library(vctrs)

vec_c(1:3, table(c(1L, 2L)))
#> Error: No common type for `..1` <integer> and `..2` <table>.
```

<sup>Created on 2020-03-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

which apparently is what breaks `cheese` on dplyr 1.0.0 rev deps: 

```
[master*] 270.8 MiB ❯ revdepcheck::revdep_details(revdep = "cheese")
══ Reverse dependency check ════════════════════════════════════ cheese 0.0.3 ══

Status: BROKEN

── Newly failing

✖ checking examples ... ERROR

── Before ──────────────────────────────────────────────────────────────────────
0 errors ✔ | 0 warnings ✔ | 0 notes ✔

── After ───────────────────────────────────────────────────────────────────────
❯ checking examples ... ERROR
  Running examples in ‘cheese-Ex.R’ failed
  The error most likely occurred in:

  > ### Name: descriptives
  > ### Title: Compute descriptive statistics on columns of a data frame.
  > ### Aliases: descriptives
  >
  > ### ** Examples
  >
  > require(tidyverse)
  Loading required package: tidyverse
  ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──
  ✔ ggplot2 3.3.0           ✔ purrr   0.3.3
  ✔ tibble  2.1.3           ✔ dplyr   0.8.99.9002
  ✔ tidyr   1.0.2           ✔ stringr 1.4.0
  ✔ readr   1.3.1           ✔ forcats 0.5.0
  ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
  ✖ dplyr::filter() masks stats::filter()
  ✖ dplyr::lag()    masks stats::lag()
  >
  > #1) Default
  > heart_disease %>%
  +     descriptives()
  Error: No common type for `length$.value` <integer> and `count$.value` <table>.
  Backtrace:
       █
    1. ├─heart_disease %>% descriptives()
    2. │ ├─base::withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
    3. │ └─base::eval(quote(`_fseq`(`_lhs`)), env, env)
    4. │   └─base::eval(quote(`_fseq`(`_lhs`)), env, env)
    5. │     └─`_fseq`(`_lhs`)
    6. │       └─magrittr::freduce(value, `_function_list`)
    7. │         ├─base::withVisible(function_list[[k]](value))
    8. │         └─function_list[[k]](value)
    9. │           └─cheese::descriptives(.)
   10. │             └─`%>%`(...)
   11. │               ├─base::withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
   12. │               └─base::eval(quote(`_fseq`(`_lhs`)), env, env)
   13. │                 └─base::eval(quote(`_fseq`(`_lhs`)), env, env)
   14. │                   └─cheese:::`_fseq`(`_lhs`)
   15. │                     └─magrittr::freduce(value, `_
  Execution halted

1 error ✖ | 0 warnings ✔ | 0 notes ✔
```

**Comment by DavisVaughan** [comment]
I am not sure if this should be allowed? @lionel-? I don't think there is a clear answer of what the common type would be here, i.e. I don't know how to mix a contingency table with an integer vector

**Comment by lionel-** [comment]
Maybe it should be integer? In some way "table" is the constrained type? I could see cases where you'd want to treat a contingency vector as integer, e.g. to incorporate the result in another vector. This seems tricky though.

**Comment by hadley** [comment]
I think it's fine to leave this as an error

**Pull request merged**